File: js/y2015/d01.js
const { fmtAnsWithRuntime } = require('../util.js');

function partOne(input) {
  let floor = 0;
  for (let a = 0; a < input.length; a++) {
    if (input[a] === "(") {
      floor++;
    }
    if (input[a] === ")") {
      floor--;
    }
  }
  return floor;
}

function partTwo(input) {
  let floor = 0;
  for (let a = 0; a < input.length; a++) {
    if (input[a] === "(") {
      floor++;
    }
    if (input[a] === ")") {
      floor--;
    }
    if (floor < 0) return a + 1;
  }
}

function soln(rawInput) {
  fmtAnsWithRuntime(
    () => partOne(rawInput),
    () => partTwo(rawInput)
  );
}

module.exports = { soln };

File: rust/src/y2015/d01.rs
use crate::util::format_soln_string;

pub fn soln(input: String) -> String {
    let mut f = 0;
    let mut b: usize = 0;

    for (i, c) in input.chars().enumerate() {
        f += match c {
            '(' => 1,
            ')' => -1,
            _ => 0,
        };
        if b == 0 && f == -1 {
            b = i + 1;
        }
    }

    return format_soln_string(f.to_string(), b.to_string());
}
